command/eval_delete.go

```diff
@@ -0,0 +1,529 @@
+	}
+	return nil
+}
+
+// handleEvalArgDelete handles deletion and evaluation which was passed via
+// it's ID as a command argument. This is the simplest route to take and
+// doesn't require filtering or batching.
+func (e *EvalDeleteCommand) handleEvalArgDelete(evalID string) (int, error) {
+	evalInfo, _, err := e.client.Evaluations().Info(evalID, nil)
+	if err != nil {
+		return 1, err
+	}
+	return e.batchDelete([]*api.Evaluation{evalInfo})
+}
+
+// handleFlagFilterDelete handles deletion of evaluations discovered using
+// filter flags. It is unknown how many will match the operator criteria so
+// this function ensure to batch lookup and delete requests into sensible
+// numbers.
+//
+// The function is recursive and will run until it has found and deleted all
+// evals which matched the criteria. It is possible for a batch of evals to be
+// deleted, but a subsequent call to fail to due temporary issues. This is fine
+// and safe can the command can be re-run if this occurs.
+func (e *EvalDeleteCommand) handleFlagFilterDelete(nextToken, filter string) (int, error) {
+
+	// Generate the query options using the passed next token and filter. The
+	// per page value is less than the total number we can include in a single
+	// delete request. This allows us to keep the maximum return object of the
+	// eval list call to fairly small.
+	//
+	// The JSON serialized evaluation API object is 350-380B in size.
+	// 2426 * 380B (3.8e-4 MB) = 0.92MB. We may want to make this configurable
+	// in the future, but this is counteracted by the CLI logic which will loop
+	// until the user tells it to exit, or all evals matching the filter are
+	// deleted. 2426 * 3 falls below the maximum limit for eval IDs in a single
+	// delete request.
+	opts := &api.QueryOptions{
+		Filter:    filter,
+		PerPage:   2426,
+		NextToken: nextToken,
+	}
+
+	var evalsToDelete []*api.Evaluation
+
+	// Iterate 3 times, so we have a maximum size of eval IDs to delete in a
+	// single go, but don't perform a large single eval listing in a single
+	// call.
+	for i := 0; i < 3; i++ {
+
+		evalList, meta, err := e.client.Evaluations().List(opts)
+		if err != nil {
+			return 1, err
+		}
+		evalsToDelete = append(evalsToDelete, evalList...)
+
+		// Store the next token no matter if it is empty or populated.
+		opts.NextToken = meta.NextToken
+
+		// If there is no next token, ensure we exit and avoid any new loops
+		// which will result in duplicate IDs.
+		if opts.NextToken == "" {
+			break
+		}
+	}
+
+	// The filter flags are operator controlled, therefore ensure we actually
+	// found some evals to delete. Otherwise, inform the operator their flags
+	// are potentially incorrect.
+	if len(evalsToDelete) == 0 {
+		if e.numDeleted > 0 {
+			return 0, nil
+		} else {
+			return 1, errors.New("failed to find any evals that matched filter criteria")
+		}
+	}
+
+	if code, err := e.batchDelete(evalsToDelete); err != nil {
+		return code, err
+	}
+
```

Comment:
would it make sense to log progress here?

Reply:
Yeah good shout, i'll add a line per iteration so things are very clear.

Reply:
This is resolved within https://github.com/hashicorp/nomad/pull/13492/commits/26e1e28dfa5290563af8a20748d865c2118da0ce along with another problem I found.